backend/routes/util.py

```diff
@@ -0,0 +1,296 @@
+    """
+    Set an HTTP-only secure cookie on the response for authentication tokens.
+
+    Args:
+        response (Response): Flask response object.
+        cookie_name (str): The name of the cookie to set.
+        value (str): The value to store in the cookie.
+    """
+    response.set_cookie(
+        cookie_name,
+        value,
+        httponly=True,
+        secure=True,
+        samesite="Strict",
+    )
+
+
+def validate_input(func):
+    """
+    Decorator function to validate the presence and format of `username`
+    and `password` fields in the JSON request data.
+
+    Args:
+        func (function): The function to be decorated.
+
+    Returns:
+        function: The wrapped function with input validation.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if not request.is_json:
+            return jsonify({"error": "Request must be JSON"}), HTTPStatus.BAD_REQUEST
+
+        try:
+            data = request.json
+        except ValueError:
+            return jsonify({"error": "Invalid JSON format"}), 400
+        if not data or "username" not in data or "password" not in data:
+            return (
+                jsonify({"error": "Username and password are required"}),
+                HTTPStatus.BAD_REQUEST,
+            )
+
+        username = data["username"]
+        password = data["password"]
+
+        # Validate username and password
+        if not (validate_username(username) and validate_password(password)):
+            return (
+                jsonify({"error": "Invalid username or password format"}),
+                HTTPStatus.BAD_REQUEST,
+            )
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def login_required(f):
+    """Decorator to enforce authentication for a route.
+
+    This decorator checks for the presence of a valid authentication token in the request
+    cookies before allowing access to the decorated route. If the token is missing or invalid,
+    it returns an unauthorized error response.
+
+    Args:
+        f (function): The function to be decorated, representing the route handler.
+
+    Returns:
+        function: The wrapped function with authentication enforcement.
+
+    Raises:
+        Unauthorized: If the token is missing, expired, or invalid.
+
+    Examples:
+        @app.route("/protected")
+        @login_required
+        def protected_route():
+            return jsonify({"message": "Access granted"})
+    """
+
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        token = request.cookies.get("id_token")  # Or use access_token if preferred
+
+        if not token:
+            return jsonify({"error": "Authorization required"}), HTTPStatus.UNAUTHORIZED
+
+        try:
+            # Decode and verify the token
+            decoded_token = decode_and_verify_token(token)
+            # You can add any additional verification logic here if needed
+            request.id_token = (
```

Comment:
Your proposed change to the login_required decorator introduces an issue: the f route handler function won't be executed because the decode_and_verify_token function is returned instead of calling the decorated route.